include/event_manager/event_manager_priv.h

```diff
@@ -0,0 +1,220 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/* There are 3 levels of priorities defining an order at which event listeners
+ * are notified about incoming events.
+ */
+
+#define _SUBS_PRIO_FIRST  0
+#define _SUBS_PRIO_NORMAL 1
+#define _SUBS_PRIO_FINAL  2
+
+
+/* Convenience macros generating section names. */
+
+#define _SUBS_PRIO_ID(level) _CONCAT(_CONCAT(_prio, level), _)
+
+#define _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio)						\
+		_CONCAT(_CONCAT(event_subscribers_, ename), prio)
+
+#define _EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)						\
+		STRINGIFY(_EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+
+/* Declare a zero-length subscriber. */
+#define _EVENT_SUBSCRIBERS_EMPTY(ename, prio)								\
+	const struct {} _CONCAT(_EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio), empty)			\
+	__attribute__((__section__(STRINGIFY(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio))))) = {};
+
+
+/* Convenience macros generating section start and stop markers. */
+
+#define _EVENT_SUBSCRIBERS_START(ename, prio)							\
+		_CONCAT(__start_, _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+#define _EVENT_SUBSCRIBERS_STOP(ename, prio)							\
+		_CONCAT(__stop_,  _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+
+#define _EVENT_SUBSCRIBERS_DECLARE(ename)										\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))[];
+
+
+/* Macro defining empty subscribers on each priority level.
+ * Each event type keeps an array of subscribers for every priority level.
+ * It can happen that for a given priority no subscriber will be registered.
+ * This macro declares zero-length subscriber that will cause required section
+ * to be generated by the linker. If no subscriber is registered at this
+ * level section will remain empty.
+ */
+#define _EVENT_SUBSCRIBERS_DEFINE(ename)					\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))	\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))	\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))
+
+
+/* Subscribe a listener to an event. */
+#define _EVENT_SUBSCRIBE(lname, ename, prio)								\
+	const struct event_subscriber _CONCAT(_CONCAT(__event_subscriber_, ename), lname) __used	\
+	__attribute__((__section__(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)))) = {			\
+		.listener = &_CONCAT(__event_listener_, lname),						\
+	}
```

Comment:
any reason not to use https://docs.zephyrproject.org/latest/kernel/iterable_sections/index.html ?

Reply:
I think I had a look during the last rework if we can use more of standard macros. There must have been something that prevented it but I don't recall what it was.